CommonTools/RecoAlgos/plugins/TrackPUIDMVAProducer.cc

```diff
@@ -0,0 +1,150 @@
+      mva_(iConfig.getParameter<edm::FileInPath>("trackPUID_mtdQualBDT_weights_file").fullPath()) {
+  produces<edm::ValueMap<float>>(mvaName);
+}
+
+// Configuration descriptions
+void TrackPUIDMVAProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("tracksSrc", edm::InputTag("generalTracks"))->setComment("Input tracks collection");
+  desc.add<edm::InputTag>("tracksMTDSrc", edm::InputTag("trackExtenderWithMTD"))
+      ->setComment("Input tracks collection for MTD extended tracks");
+  desc.add<edm::InputTag>("btlMatchChi2Src", edm::InputTag("trackExtenderWithMTD", "btlMatchChi2"))
+      ->setComment("BTL Chi2 Matching value Map");
+  desc.add<edm::InputTag>("btlMatchTimeChi2Src", edm::InputTag("trackExtenderWithMTD", "btlMatchTimeChi2"))
+      ->setComment("BTL Chi2 Matching value Map");
+  desc.add<edm::InputTag>("etlMatchChi2Src", edm::InputTag("trackExtenderWithMTD", "etlMatchChi2"))
+      ->setComment("ETL Chi2 Matching value Map");
+  desc.add<edm::InputTag>("etlMatchTimeChi2Src", edm::InputTag("trackExtenderWithMTD", "etlMatchTimeChi2"))
+      ->setComment("ETL Chi2 Matching value Map");
+  desc.add<edm::InputTag>("mtdTimeSrc", edm::InputTag("trackExtenderWithMTD", "tmtd"))
+      ->setComment("MTD TIme value Map");
+  desc.add<edm::InputTag>("pathLengthSrc", edm::InputTag("trackExtenderWithMTD", "pathLength"))
+      ->setComment("MTD PathLength value Map");
+  desc.add<edm::FileInPath>("trackPUID_mtdQualBDT_weights_file",
+                            edm::FileInPath("CommonTools/RecoAlgos/data/clf4D_MTDquality_bo.xml.gz"))
+      ->setComment("Track PUID 4D BDT weights");
+  descriptions.add("trackPUIDMVAProducer", desc);
+}
+
+template <class H, class T>
+void TrackPUIDMVAProducer::fillValueMap(edm::Event& iEvent,
+                                        const edm::Handle<H>& handle,
+                                        const std::vector<T>& vec,
+                                        const std::string& name) const {
+  auto out = std::make_unique<edm::ValueMap<T>>();
+  typename edm::ValueMap<T>::Filler filler(*out);
+  filler.insert(handle, vec.begin(), vec.end());
+  filler.fill();
+  iEvent.put(std::move(out), name);
+}
+
+void TrackPUIDMVAProducer::produce(edm::Event& ev, const edm::EventSetup& es) {
+  edm::Handle<reco::TrackCollection> tracksH;
+  ev.getByToken(tracksToken_, tracksH);
+  const auto& tracks = *tracksH;
+
+  edm::Handle<reco::TrackCollection> tracksMTDH;
+  ev.getByToken(tracksMTDToken_, tracksMTDH);
+
+  edm::Handle<edm::ValueMap<float>> btlMatchChi2H;
+  edm::Handle<edm::ValueMap<float>> btlMatchTimeChi2H;
+  edm::Handle<edm::ValueMap<float>> etlMatchChi2H;
+  edm::Handle<edm::ValueMap<float>> etlMatchTimeChi2H;
+  edm::Handle<edm::ValueMap<float>> mtdTimeH;
+  edm::Handle<edm::ValueMap<float>> pathLengthH;
+
+  ev.getByToken(btlMatchChi2Token_, btlMatchChi2H);
+  auto btlMatchChi2 = *btlMatchChi2H.product();
```

Comment:
```c++
   edm::Handle<edm::ValueMap<float>> btlMatchChi2H;
   ev.getByToken(btlMatchChi2Token_, btlMatchChi2H);
   auto btlMatchChi2 = *btlMatchChi2H.product();
```
can be replaced by
```c++
auto btlMatchChi2 = ev.get(btlMatchChi2Token_);
```

Reply:
Better be
```cpp
const auto& btlMatchChi2 = ev.get(btlMatchChi2Token_);
```
(unless the copying is intentional)